doc/dev/error-categorization.md

```diff
@@ -0,0 +1,173 @@
+# Error Categorization
+
+This document explains how to categorize errors and should act as a form of
+guideline. The categorization aims to reduce duplicated errors and the
+maintenance effort for errors.
+
+## Categorization Mindset
+
+Errors along with their unique code only primarily exist because they can be
+**reacted differently on programmatically**. To get an idea of programmatic
+reactions take a method call which returns a `Timeout` error. Naturally a
+senseful reaction would be to retry at a later point in time. So the reaction
+here would be to retry with a time based approach. On the other hand it does not
+make sense to differentiate between `No Write Permission` and `No Read Permission`
+as the application just knows that it simply cannot access the desired resource
+and tells the user to grant it.
+
+Categories are leaf based, so you cannot put an error into a node (branch) such
+as `Permanent errors` (see below).
+
+If you feel for a new category, please forge a design decision document and make
+a PR to Elektra's repo.
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an
+error/warning in there.
+
+For a complete structural overview please visit the corresponding [design
+decision document](../decisions/error_codes.md)
+
+### Permanent errors ("C01000")
+
+The branch category `Permanent Errors` refer to such errors which cannot be
+fixed by retry at all. `Permanent Errors` are subdivided into
+
+- Resource
+- Installation
+- Logical
+
+#### Resource ("C01100")
+
+`Resource Errors` as a branch category are all kinds of errors which are either
+permission related, existence related or missing resources such as memory, RAM,
+etc. Examples are missing files, insufficient permissions or out of memory. This
+category forces users or applications to provide additional resources, create a
+missing file/ directory etc or to provide the relevant permissions.
+
+##### Memory Allocation ("C01110")
+
+`Memory Allocation Errors` are errors which come from failed `elektraMalloc`
+calls primarily as no more memory could be allocated for the application. Errors
+with not enough hard disc space do not belong here but into `General Resource`.
+Such errors will gain special handling in future releases and users cannot deal
+with such errors as of now.
+
+##### General Resource ("C01120")
```

Comment:
Alright, since Elektra is your project and more or less belongs to you I will also incorporate branches as categories despite it is really against my desired option.

Reply:
We are searching for the best option, not for a personal opinion. See above: the user view was not clear to me, thus I did not see how the extension will work.

If you find a good name for "General Resource", I'll happily accept your proposal.

Reply:
Unfortunately there are many good options and there is no clear "best" option. So it boils down to some sort of personal opinion in the end. You want selectable branch categories because it is sometimes easier to categorize and name whereas I want leafs only because I assume that developers will be too lazy to put errors correctly (it honestly is a boring task). Both has its pros and cons and you insisted multiple times on branch categories so it is ok for me.

> If you find a good name for "General Resource", I'll happily accept your proposal.

No because there is no self explaining name for "Non-memory allocation failed errors but resource related errors" as far as I know.

Reply:
> Both have its pros and cons and you insisted multiple times on branch categories so it is ok for me.

For such situations we should write design decisions. The 3rd way is that there is some way to specify which categories can have errors and which are too generic (It seems like "Resource" seems to be the only branch that should contain errors).

> No because there is no self explaining name for "Non-memory allocation failed errors but resource related errors" as far as I know.

Yes, that is why I thought would make sense to directly have errors in "Resource". Because the "General Resource" is actually "Resource" \ "Memory Errors" (\ as "set difference").